### PR TITLE
Fix loop when improvement is unbuildable and removes features

### LINKED ISSUE #11332: StackOverflowError when building on unbuildable terrain (customs house on snow)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Game Version

4.10.20

### Describe the bug

While playing a single-player game, the game suddenly crashed after building a customs house on a snow tile.

### Steps to Reproduce

1. Get a great person.
2. Move to a snow tile.
3. Create the great improvement.

I still have to thoroughly test this, but it was the first occurrence of such an event.

### Screenshots

_No response_

### Link to save file

https://gist.github.com/Bjastkuliar/ce292a30dbbd168341e6e7ff28443786

### Operating System

Windows

### Additional Information

I've just updated the game today and therefore it might be an issue of this specific version, I just downloaded the JAR file and swapped the old JAR in the unciv folder (I originally had downloaded the whole "windows edition", but I was using the JRE game, not the .exe

### PR DESCRIPTION
**Pull request by SeventhM**
Resolves #11332

### CODE REVIEW

File: core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt [hunk 1]
```diff
@@ -110,13 +110,14 @@ class TileInfoImprovementFunctions(val tile: Tile) {
...
             if (!hasUnique(UniqueType.RemovesFeaturesIfBuilt, stateForConditionals)) return false
-            if (knownFeatureRemovals == null) return false
-            val featureRemovals = tile.terrainFeatures.map {
-                feature -> tile.ruleset.tileRemovals.firstOrNull{ it.name == Constants.remove + feature } }
-            if (featureRemovals.any { it != null && it !in knownFeatureRemovals }) return false
+            if (knownFeatureRemovals.isNullOrEmpty()) return false
+            val featureRemovals = tile.terrainFeatures.mapNotNull { feature -> 
+                tile.ruleset.tileRemovals.firstOrNull { it.name == Constants.remove + feature } }
+            if (featureRemovals.isEmpty()) return false
+            if (featureRemovals.any { it !in knownFeatureRemovals }) return false
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt:117]: What does it mean, if a feature removal is not "known"? It's in the ruleset, what are we actually checking here?

Comment (SeventhM) [reply core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt:117]: Which removals have we unlocked the techs for. Definition is further up; line 82

Comment (SeventhM) [reply core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt:117]: The check here is "of all of the removals that could apply, do we have all of their techs unlocked?" I would go "huh, wait, why don't we check that directly?" but it was like that before. I would assume there's a reason... Hopefully

**Review by yairm210 — APPROVED** [review]

File: core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt (hunk 1 above)
Comment (yairm210) [reply core/src/com/unciv/logic/map/tile/TileInfoImprovementFunctions.kt:117]: Looks like this whole thing requires cleanup, not for this pr

**Pull request merged**